singularity-eos/eos/singularity_eos.cpp

```diff
@@ -343,327 +350,575 @@ int get_sg_eos( // sizing information
-        for (int m = 0; m < nmat; ++m)
-          mass_sum += frac_mass_v(i, m);
-        // validate specific volume
-        spvol_v(i) = r_is_inp ? spvol_v(i) : vol_v(i) / mass_sum;
-        // calculate specific internal energy if not already known
-        if (!s_is_inp) {
-          Real eng_sum{0.0};
-          Real vol_sum{0.0};
+  const std::string perf_nums =
+      "[" + std::to_string(nmat) + "," + std::to_string(ncell) + "]";
+  switch (input_int) {
+  case -3: {
+    // T-rho input
+    // set frac_vol = 1/nmat
+    // set rho_i = nmat / spvol * frac_mass_i
+    // iterate PTE solver to obtain internal energies
+    // that results in the input T
+    pte_solver_scratch_size = PTESolverFixedTRequiredScratch(nmat);
+    solver_scratch = ScratchV<double>(VAWI("PTE::scratch solver"), scratch_size,
+                                      pte_solver_scratch_size);
+    const std::string rt_name = "PTE::solve (rho,T) input" + perf_nums;
+    portableFor(
+        rt_name.c_str(), 0, ncell, PORTABLE_LAMBDA(const int &iloop) {
+          // cell offset
+          const int i{offsets_v(iloop) - 1};
+          // get "thread-id" like thing with optimization
+          // for small loops
+          const int32_t token{tokens.acquire()};
+          const int32_t tid{small_loop ? iloop : token};
+          double mass_sum{0.0};
+          // normalize mass fractions
+          // first find the mass sum
+          // also set idxs as the decrement of the eos offsets
+          // to take into account 1 based indexing in fortran
+          for (int m = 0; m < nmat; ++m) {
+            mass_sum += frac_mass_v(i, m);
+            pte_idxs(tid, m) = eos_offsets_v(m) - 1;
+            frac_vol_v(i, m) = 0.0;
+          }
+          for (int m = 0; m < nmat; ++m) {
+            frac_mass_v(i, m) /= mass_sum;
+          }
+          // set inputs
+          int npte = 0;
           for (int m = 0; m < nmat; ++m) {
-            Real rho = 1.0 / spvol_v(i);
-            Real T = temp_v(i) * ev2k;
-            Real press_mat = press_v(i);
-            Real spie = sie_v(i);
-            Real cv_mat, bmod_mat;
-            int mid = eos_offsets_v(m) - 1;
-            eos_v(mid).FillEos(rho, T, spie, press_mat, cv_mat, bmod_mat, output, cache);
-            eng_sum += spie * frac_mass_v(i, m);
-            vol_sum += frac_mass_v(i, m) / rho;
-            if (vol_sum < 0.0)
-              printf("Neg. Volume:\ninput: %i mat: %i r: %e t: %e s: %e p: %e\n",
-                     input_int, mid + 1, rho, T, spie, press_mat);
-            // if (!r_is_inp) printf("mat: %i rho lookup: %e\n", eos_offsets_v(m), rho);
+            if (frac_mass_v(i, m) > 1.e-12) {
+              pte_idxs(tid, npte) = eos_offsets_v(m) - 1;
+              pte_mats(tid, npte) = m;
+              npte += 1;
+            }
+            vfrac_pte(tid, m) = 0.0;
+            sie_pte(tid, m) = 0.0;
+            temp_pte(tid, m) = 0.0;
+            press_pte(tid, m) = 0.0;
+          }
```

Comment:
Similarly, this can be put in a function too I think

Reply:
Before we get too carried away with code making everything a function. There is probably a much better way of handling this entire function as a whole and we have discussed that very thing and I just don't think this is the PR to do that. First correctness, then we can think more about what the correct function is and how to go about reducing duplication.

Reply:
I'd rather keep things clean here---though I'm willing to do a cleanup PR after this one, assuming that cleanup actually happens.

Reply:
My main motivation in bringing this up is because I was trying to read through the code to help in debugging the xRAGE integration. I was thinking that making the code more DRY would aid in the debugging process. In speaking from my own experience, I've added entirely too many bugs by not putting code in a function when I wanted to reuse it.

At this point in the debugging process, I think you've eliminated this code from being the source of the error we're tracking down, but I think it may be helpful in the future for readability and future debugging.

Reply:
done